### PR TITLE
Add new tests for v200 schema

### PR DESCRIPTION
**Pull request by mmulholla**

### CODE REVIEW

**Comment by mmulholla** [comment]
Interestingly the new tests are run automatically as part of the pull request build.

File: test/devfiles/v200/commands/ExecMissingEnvOrder.yaml
Comment (amisevsk) [inline test/devfiles/v200/commands/ExecMissingEnvOrder.yaml:15]: I'm not sure what this test is checking.

File: test/go/src/tests/api-v200_test.go
Comment (amisevsk) [inline test/go/src/tests/api-v200_test.go:15]: What do we mean by "users struct"?

**Review by amisevsk — COMMENTED** [review]
Please separate out testing code and readme changes from `yaml` files used for test cases; it's quite hard to review this PR as-is (it's a flat list of 141 files basically)

I'm really not clear on what we're actually trying to test here -- are we trying to validate that the schema we're generating matches what's expected? Are we actually just testing that the json generator does what it says on the box? If we have doubts about the correctness of the schema generator, we should be testing the schema generator.

The structure of the testing code is also very strange to me -- it seems vastly overcomplicated for what would be a fairly simple task. Have we considered using a [fuzzer](https://github.com/google/gofuzz) to generate randomly filled structs, serialize them to json, and check them against the schema?

File: test/devfiles/v200/commands/ExecMissingEnvOrder.yaml
Comment (mmulholla) [reply test/devfiles/v200/commands/ExecMissingEnvOrder.yaml:15]: May be a redendant one, just checking it does not matter which order you put the name and value in.

File: test/go/src/tests/api-v200_test.go
Comment (mmulholla) [reply test/go/src/tests/api-v200_test.go:15]: oops, cut and paste error, will fix it.

File: test/devfiles/v200/commands/ExecMissingEnvOrder.yaml
Comment (mmulholla) [reply test/devfiles/v200/commands/ExecMissingEnvOrder.yaml]: I fixed the issues you raised and also one I noticed. The easiest way to review is to go to the branch and then the test subdirectory which the PR creates: https://github.com/mmulholla/api/tree/AddAPITests/test

There you can see the README and easily browse the new files.

The tests may seem a bit pointless right now, but will be useful in the future when the schema is updated and we check if v200 devfiles still work with the new schema.

**Comment by mmulholla** [comment]
I noticed an error in the test logic which was giving some false positives. Going to cancel this pull request until I have addressed the issues

**Comment by mmulholla** [comment]
Fixed the issues and added a commit to the pull request so have re-opened it.

**Comment by mmulholla** [comment]
Will close the pull request again and re-open later once the endpoint proticol issue has concluded and test modified accordingly.

File: test/devfiles/v200/commands/ExecMissingEnvOrder.yaml
Comment (mmulholla) [reply test/devfiles/v200/commands/ExecMissingEnvOrder.yaml]: Also not sure how these tests are over complicated. They perform two goals :
 
1. test every attriibute in the schema, make sure mandatory attributes are indeed mandatory and optional ones are indeed optional. Then where attributes have a restriction make sure those restrictions are enforced. 
2. Make it easy to add new tests for a new attribute or different schema.

Add to this that the validator only reports the first problem, so you need a test for every error case. There is obviously a bunch of different ways you can achieve these goals but, would be equally complicated.

Adding a fuzzer to generate values would be interesting but would need all the logic in the go code which then makes it more difficult to add and changes tests. Also not sure how it works when an attribute is restricted to a range of values.

Maybe i misunderstand how the fuzzer would be used.

File: test/devfiles/v200/commands/ExecBadDefault.yaml
Comment (sleshchenko) [inline test/devfiles/v200/commands/ExecBadDefault.yaml:1]: filename is too generic. At the first look, I thought it's about default value for some field, but it's about isDefault field. So, maybe ExecBadDefaultInGroup would be a bit more self-descriptive?

File: test/devfiles/v200/commands/ExecBadDefault.yaml
Comment (sleshchenko) [inline test/devfiles/v200/commands/ExecBadDefault.yaml:18]: Could you please add newlines in files?

File: test/devfiles/v200/commands/ExecBadDefault.yaml
Comment (mmulholla) [reply test/devfiles/v200/commands/ExecBadDefault.yaml:1]: OK, will rename them.

File: test/devfiles/v200/commands/ExecBadDefault.yaml
Comment (mmulholla) [reply test/devfiles/v200/commands/ExecBadDefault.yaml:18]: Sorry, not sure what you mean.

Comment (sleshchenko) [reply test/devfiles/v200/commands/ExecBadDefault.yaml:18]: I mean this 
Screenshot_20201111_171812

here you can find why it's better to add it https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline/729725

Comment (mmulholla) [reply test/devfiles/v200/commands/ExecBadDefault.yaml]: Thank you  all done:
- added newlines
- updated tests for recent changes - attributes updates and name and id restrictions
- added tests for. the new name and id restrictions 

Read for review again, hopefully all good this time :-)

Comment (mmulholla) [reply test/devfiles/v200/commands/ExecBadDefault.yaml]: Some tests failing on the build. failing to find files - need to dig into it, all was good locally on my mac.

File: test/go/src/tests/api-v200_test.go
Comment (mmulholla) [reply test/go/src/tests/api-v200_test.go]: all good now, differenence in how Mac and linux handle file names.

File: test/devfiles/v200/commands/ExecBadDefault.yaml
Comment (mmulholla) [reply test/devfiles/v200/commands/ExecBadDefault.yaml]: All good now, difference in how Mac and linux handle case in file names.

File: test/go/src/tests/api-v200_test.go
Comment (sleshchenko) [inline test/go/src/tests/api-v200_test.go:214]: It's also useful to see number of skipped tests

File: test/go/src/tests/api-v200_test.go
Comment (sleshchenko) [inline test/go/src/tests/api-v200_test.go:237]: What does this test do?

**Review by sleshchenko — COMMENTED** [review]
The test code seems to suppress a lot of errors and is not really readable.
I propose to move forward and improve it a bit like the following example https://github.com/mmulholla/api/compare/AddAPITests...sleshchenko:AddAPITestsPatch?expand=1
^ it's not finished yet but I hope shows the idea

**Comment by mmulholla** [comment]
@sleshchenko - please let me know when you consider your branch complete

File: test/README.md
Comment (amisevsk) [reply test/README.md:9]: Remove the extra 't' while we're at it :)

**Review by amisevsk — COMMENTED** [review]
I won't block merging or anything, but in general, I'm not a fan of tests in this style -- we're basically testing that a few annotations are set correctly in `pkg/apis`. However, since the tests are written manually, there's a much larger chance that there is an error in the tests than the code itself. 

I don't see a case where I could make an error writing code that is caught by these tests -- they basically will fail whenever a mandatory/optional/enum field is changed, but the solution there is to just update the tests to accommodate the change. It feels like it'll end up being rote maintenance for tests for the sake of having them. 

Also, these tests will have to be kept in sync with documentation, which I suppose is the ultimate source of truth. This seems like it will be a large maintenance burden for little benefit.

**Comment by elsony** [comment]
@amisevsk The main goal of these tests is to make sure that we don't change the API and breaking the existing devfiles. In case there is an error occurs, it is either some changes have been introduced that breaks existing devfiles. It is true that the document should match but there is no way for us to confirm that the doc matches the code to enforce compatibility. These tests are meant for catching that. For intentional changes on the schema, I'd expect the test change will go with the code changes to reflect those.

File: test/README.md
Comment (mmulholla) [reply test/README.md]: fixed

File: test/README.md
Comment (mmulholla) [reply test/README.md]: fixed

File: test/README.md
Comment (mmulholla) [reply test/README.md:41]: Fixed.

File: test/README.md
Comment (mmulholla) [reply test/README.md:39]: Fixed

File: test/README.md
Comment (mmulholla) [reply test/README.md]: Typos fixed. I am planning to add a link once this is merged so that it will link into the devfile/api repo.

File: test/README.md
Comment (mmulholla) [reply test/README.md]: Fixed.

File: test/go/src/tests/api-v200_test.go
Comment (mmulholla) [reply test/go/src/tests/api-v200_test.go]: Good spot, updated Fatalf to Errorf.

File: test/go/src/tests/api-v200_test.go
Comment (mmulholla) [reply test/go/src/tests/api-v200_test.go]: Removed - just a harmless old function I used to try stuff out.

**Comment by mmulholla** [comment]
Further to Angels comments about complexity I simplified the directory structure and renamed the test to make thing more obvious. Also I set it up so when we want to create a new set of tests for a new version of the schema it is easy to re-use this test directory structure as a starting point - I added instruction for this in the README.

File: test/go/src/tests/api-v200_test.go
Comment (mmulholla) [reply test/go/src/tests/api-v200_test.go]: Sure. will also address this is a subsequent pull request.

File: test/go/src/tests/api-v200_test.go
Comment (mmulholla) [reply test/go/src/tests/api-v200_test.go]: Agree it is a bit cumbersome. I saw your branch and plan to handle this in a later pull request. Not decided yet but thinking I may create, for example,  a new json-utils.go file to provide an api for access to the json files. 

Opened this issue: https://github.com/devfile/api/issues/298

**Review by elsony — APPROVED** [review]
/lgtm (with some comments going to be covered by #298)

**Pull request merged**
